appshell/appshell_extensions_mac.mm

```diff
@@ -408,23 +430,48 @@ int32 ShowOpenDialog(bool allowMulitpleSelection,
     
     [openPanel setAllowedFileTypes:allowedFileTypes];
     
-    [openPanel beginSheetModalForWindow:[NSApp mainWindow] completionHandler:nil];
-    if ([openPanel runModal] == NSOKButton)
+    // cache the browser details, so has to access
+    // the same inside the completionHandler block.
+    CefBrowserDetails browserDetails;
+    browserDetails.m_browser  = browser;
+    browserDetails.m_response = response;
+    
+    s_panel_to_browser[openPanel] = browserDetails;
+    
+    [openPanel beginSheetModalForWindow:[NSApp mainWindow] completionHandler: ^(NSInteger returnCode)
     {
-        NSArray* urls = [openPanel URLs];
-        for (NSUInteger i = 0; i < [urls count]; i++) {
-            selectedFiles->SetString(i, [[[urls objectAtIndex:i] path] UTF8String]);
+        panel_iterator _browserDetails;
+        _browserDetails = s_panel_to_browser.find(openPanel);
+        if(_browserDetails != s_panel_to_browser.end()){
+            
+            CefRefPtr<CefProcessMessage> _response = _browserDetails->second.m_response;
+            CefRefPtr<CefBrowser>        _browser  = _browserDetails->second.m_browser;
+            
+            if(_response && _browser){
+
+                NSArray *urls = [openPanel URLs];
+                CefRefPtr<CefListValue> selectedFiles = CefListValue::Create();
+                if (returnCode == NSModalResponseOK){
+                    for (NSUInteger i = 0; i < [urls count]; i++) {
```

Comment:
We could use, but this piece of code was just borrowed from older way of doing this. Do you see any advantages of using `for(id url in url;s)`?